src/hash.cpp

```diff
@@ -24,33 +24,22 @@ namespace mococrw
 {
 using namespace openssl;
 
-static const EVP_MD* _getDigestFn (DigestTypes digestType) {
-    switch (digestType) {
-        case DigestTypes::SHA256:
-            return lib::OpenSSLLib::SSL_EVP_sha256();
-        case DigestTypes::SHA512:
-            return lib::OpenSSLLib::SSL_EVP_sha512();
-        default:
-            throw std::runtime_error("Unknown digest type");
-    }
-}
-
-static EVP_MD_CTX _initCtx (DigestTypes digestType) {
-    const EVP_MD* digestFn = _getDigestFn(digestType);
-    EVP_MD_CTX digestCtx;
-    EVP_MD_CTX_init(&digestCtx);
-    EVP_DigestInit_ex(&digestCtx, digestFn, NULL);
+static SSL_EVP_MD_CTX_Ptr _initCtx (DigestTypes digestType) {
+    const SSL_EVP_MD_Ptr digestFn = _getMDPtrFromDigestType(digestType);
+    SSL_EVP_MD_CTX_Ptr digestCtx = _EVP_MD_CTX_create();
+    lib::OpenSSLLib::SSL_EVP_MD_CTX_init(digestCtx.get());
+    lib::OpenSSLLib::SSL_EVP_DigestInit_ex(digestCtx.get(), digestFn.get(), NULL);
     return digestCtx;
 }
 
-static void _finalizeDigest (EVP_MD_CTX& digestCtx, uint8_t* outDigest, unsigned int &outLength) {
-    EVP_DigestFinal_ex(&digestCtx, outDigest, &outLength);
-    EVP_MD_CTX_cleanup(&digestCtx);
+static void _finalizeDigest (SSL_EVP_MD_CTX_Ptr &digestCtx, uint8_t* outDigest, unsigned int &outLength) {
+    lib::OpenSSLLib::SSL_EVP_DigestFinal_ex(digestCtx.get(), outDigest, &outLength);
```

Comment:
I guess we need to talk about this again. This is not the style of the library and also does not do the error handling that we would like to see here.  The high level code should not call openssl functions directly but call abstractions from openssl_wrap that do the error handling for you. This piece of code has the effect now that error transferred via return value of the openssl functions are (silently) not handled.